app/code/Magento/Catalog/Block/Product/View/Attributes.php

```diff
@@ -115,6 +115,7 @@ protected function isVisibleOnFrontend(
         \Magento\Eav\Model\Entity\Attribute\AbstractAttribute $attribute,
         array $excludeAttr
     ) {
-        return ($attribute->getIsVisibleOnFront() && !in_array($attribute->getAttributeCode(), $excludeAttr));
+        return ($attribute->getIsVisibleOnFront() && !in_array($attribute->getAttributeCode(), $excludeAttr) &&
+            (!empty($attribute->getApplyTo()) && in_array($this->getProduct()->getTypeId(), $attribute->getApplyTo())));
```

Comment:
This gets too messy, extract these conditions to separate private method explaining what it actually does - for example:
`$attribute->isVisibleOnFront() && !$this->isAttributeExcluded($attribute) && $this->isAttributeVisibleForProduct($attribute, $this->getProduct())`